File: android/app/src/main/java/com/fitrater/app/ui/screens/PieceDetailSheet.kt
package com.fitrater.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.FilterQuality
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.fitrater.app.data.model.ClosetItem
import com.fitrater.app.data.repo.Repo
import com.fitrater.app.ui.theme.Eyebrow
import com.fitrater.app.ui.theme.HemColors
import com.fitrater.app.ui.theme.HemSpace
import com.fitrater.app.ui.theme.HemType
import com.fitrater.app.ui.theme.SerifDisplay

/**
 * Piece detail bottom sheet — shows a Studio piece with iteration history and edit CTA.
 * Extracted from StudioScreen so it can be reused from Home's "Latest" card.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PieceDetailSheet(
    item: ClosetItem,
    imageUrl: String?,
    onDismiss: () -> Unit,
    onEdit: () -> Unit,
) {
    var chain by remember(item.id) { mutableStateOf<List<ClosetItem>>(emptyList()) }
    var chainUrls by remember(item.id) { mutableStateOf<Map<String, String>>(emptyMap()) }
    var current by remember(item.id) { mutableStateOf(item) }
    var wornCount by remember(item.id) { mutableStateOf(0) }

    LaunchedEffect(current.id) {
        val id = current.id ?: return@LaunchedEffect
        wornCount = runCatching { Repo.tryOnCountForPiece(id) }.getOrDefault(0)
    }

    LaunchedEffect(current.id) {
        val list = mutableListOf<ClosetItem>()
        var cursor: ClosetItem? = current
        val visited = mutableSetOf<String>()
        while (cursor?.parent_id != null && cursor.parent_id !in visited) {
            visited.add(cursor.parent_id!!)
            val p = runCatching { Repo.closetItemById(cursor!!.parent_id!!) }.getOrNull() ?: break
            list.add(0, p)
            cursor = p
        }
        val kids = current.id?.let { runCatching { Repo.closetChildren(it) }.getOrDefault(emptyList()) }.orEmpty()
        list.addAll(kids)
        chain = list
        val urls = mutableMapOf<String, String>()
        list.forEach { c ->
            val id = c.id ?: return@forEach
            val direct = c.image_url
            if (!direct.isNullOrBlank()) { urls[id] = direct; return@forEach }
            val path = c.image_path ?: return@forEach
            runCatching { Repo.signedClosetUrl(path) }.getOrNull()?.let { urls[id] = it }
        }
        chainUrls = urls
    }
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        containerColor = HemColors.Paper,
    ) {
        Column(
            Modifier
                .fillMaxWidth()
                .padding(horizontal = HemSpace.gutter, vertical = HemSpace.md),
        ) {
            val isImported = current.image_path?.startsWith("references/") == true
            val effectiveUrl = current.id?.let { chainUrls[it] } ?: imageUrl
            Row(verticalAlignment = Alignment.CenterVertically) {
                Box(
                    Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .background(if (isImported) HemColors.CardCream else HemColors.Ink.copy(alpha = 0.08f))
                        .border(1.dp, HemColors.Hairline, RoundedCornerShape(999.dp))
                        .padding(horizontal = 10.dp, vertical = 4.dp),
                ) {
                    Text(
                        if (isImported) "IMPORTED" else "GENERATED",
                        style = HemType.bodyMuted.copy(
                            fontSize = 10.sp,
                            fontWeight = FontWeight.SemiBold,
                            letterSpacing = 1.sp,
                        ),
                    )
                }
            }
            Spacer(Modifier.height(HemSpace.sm))
            Box(
                Modifier
                    .fillMaxWidth()
                    .aspectRatio(1f)
                    .clip(RoundedCornerShape(16.dp))
                    .background(HemColors.CardCream),
                contentAlignment = Alignment.Center,
            ) {
                if (!effectiveUrl.isNullOrBlank()) {
                    AsyncImage(
                        model = effectiveUrl,
                        contentDescription = null,
                        contentScale = ContentScale.Fit,
                        filterQuality = FilterQuality.High,
                        modifier = Modifier.fillMaxSize(),
                    )
                } else {
                    Text("👕", style = HemType.serifDisplay.copy(fontSize = 96.sp, color = HemColors.Muted))
                }
            }
            Spacer(Modifier.height(HemSpace.md))
            SerifDisplay(current.name ?: "Piece")
            Spacer(Modifier.height(HemSpace.xs))
            val meta = listOfNotNull(current.subcategory ?: current.category, current.color_hex ?: current.color).joinToString(" · ")
            if (meta.isNotBlank()) Text(meta, style = HemType.bodyMuted)

            if (wornCount > 0) {
                Spacer(Modifier.height(HemSpace.xs))
                Text(
                    "WORN $wornCount TIME${if (wornCount == 1) "" else "S"}",
                    style = HemType.eyebrow.copy(color = HemColors.Bronze, letterSpacing = 2.sp, fontSize = 10.sp),
                )
            }

            if (chain.isNotEmpty()) {
                Spacer(Modifier.height(HemSpace.lg))
                Eyebrow("HISTORY · ${chain.size + 1}")
                Spacer(Modifier.height(HemSpace.sm))
                Row(
                    Modifier.horizontalScroll(rememberScrollState()),
                    horizontalArrangement = Arrangement.spacedBy(HemSpace.xs),
                ) {
                    (chain + current).forEach { c ->
                        val isCurrent = c.id == current.id
                        val u = c.id?.let { chainUrls[it] } ?: c.image_url
                        Box(
                            Modifier
                                .size(56.dp)
                                .clip(RoundedCornerShape(8.dp))
                                .background(HemColors.CardCream)
                                .border(
                                    if (isCurrent) 2.dp else 1.dp,
                                    if (isCurrent) HemColors.Bronze else HemColors.Hairline,
                                    RoundedCornerShape(8.dp),
                                )
                                .clickable {
                                    if (!isCurrent) {
                                        current = c
                                    }
                                },
                        ) {
                            if (!u.isNullOrBlank()) {
                                AsyncImage(
                                    model = u,
                                    contentDescription = null,
                                    contentScale = ContentScale.Crop,
                                    modifier = Modifier.fillMaxSize(),
                                )
                            }
                        }
                    }
                }
            }
            Spacer(Modifier.height(HemSpace.lg))
            Box(
                Modifier
                    .fillMaxWidth()
                    .height(52.dp)
                    .clip(RoundedCornerShape(999.dp))
                    .background(HemColors.Ink)
                    .clickable(onClick = onEdit),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    "✎ Edit piece",
                    style = HemType.body.copy(
                        color = Color.White,
                        fontWeight = FontWeight.SemiBold,
                    ),
                )
            }
            Spacer(Modifier.height(HemSpace.xl))
        }
    }
}

File: ios/FitScore/FitScore/Studio/PieceDetailSheet.swift
import SwiftUI

/// Piece detail bottom sheet — ports Android `PieceDetailSheet.kt`.
/// Shows the hero image, generated/imported pill, meta, history chain and
/// "WORN N TIMES" bronze eyebrow when the piece has associated try-ons.
struct PieceDetailSheet: View {
    let item: ClosetItem
    let imageUrl: String?
    let onEdit: (_ referenceUrl: String?) -> Void
    let onDismiss: () -> Void

    @State private var current: ClosetItem
    @State private var chain: [ClosetItem] = []
    @State private var chainUrls: [String: String] = [:]
    @State private var wornCount: Int = 0

    init(item: ClosetItem, imageUrl: String?, onEdit: @escaping (_ referenceUrl: String?) -> Void, onDismiss: @escaping () -> Void) {
        self.item = item
        self.imageUrl = imageUrl
        self.onEdit = onEdit
        self.onDismiss = onDismiss
        self._current = State(initialValue: item)
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                pill
                    .padding(.top, 12)
                hero
                Text(current.name ?? "Piece")
                    .font(Serif.display(24))
                    .foregroundStyle(Palette.ink)
                let meta = [current.subcategory ?? current.category, current.color_hex ?? current.color]
                    .compactMap { $0 }
                    .filter { !$0.isEmpty }
                    .joined(separator: " · ")
                if !meta.isEmpty {
                    Text(meta)
                        .font(Serif.body(13))
                        .foregroundStyle(Palette.muted)
                }
                if wornCount > 0 {
                    Eyebrow(text: "WORN \(wornCount) TIME\(wornCount == 1 ? "" : "S")")
                        .padding(.top, 4)
                }
                if !chain.isEmpty {
                    Eyebrow(text: "HISTORY · \(chain.count + 1)")
                        .padding(.top, 16)
                    ScrollView(.horizontal, showsIndicators: false) {
                        HStack(spacing: 8) {
                            ForEach(chain + [current], id: \.id) { c in
                                historyTile(c)
                            }
                        }
                    }
                }
                Button {
                    onEdit(current.id.flatMap { chainUrls[$0] } ?? imageUrl)
                } label: {
                    HStack(spacing: 8) {
                        Text("✎").foregroundStyle(.white)
                        Text("Edit piece")
                            .font(.system(size: 15, weight: .semibold))
                            .foregroundStyle(.white)
                    }
                    .frame(maxWidth: .infinity)
                    .frame(height: 52)
                    .background(Palette.ink)
                    .clipShape(Capsule())
                }
                .padding(.top, 20)
                .buttonStyle(.plain)
            }
            .padding(.horizontal, 20)
            .padding(.bottom, 32)
        }
        .background(Palette.paper.ignoresSafeArea())
        .task(id: current.id) {
            await loadContext()
        }
    }

    private var pill: some View {
        let isImported = (current.image_path ?? "").contains("references/")
        return Text(isImported ? "IMPORTED" : "GENERATED")
            .font(.system(size: 10, weight: .semibold))
            .tracking(1)
            .foregroundStyle(Palette.ink)
            .padding(.horizontal, 10).padding(.vertical, 4)
            .background(isImported ? Palette.card : Palette.ink.opacity(0.08))
            .overlay(Capsule().stroke(Palette.hairline, lineWidth: 1))
            .clipShape(Capsule())
    }

    private var hero: some View {
        let effective = current.id.flatMap { chainUrls[$0] } ?? imageUrl
        return ZStack {
            Palette.card
            if let url = effective, let u = URL(string: url) {
                AsyncImage(url: u) { phase in
                    switch phase {
                    case .success(let img): img.resizable().scaledToFit()
                    default: ProgressView().tint(Palette.muted)
                    }
                }
            } else {
                Text("👕").font(.system(size: 96)).foregroundStyle(Palette.muted)
            }
        }
        .aspectRatio(1, contentMode: .fit)
        .frame(maxWidth: .infinity)
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }

    private func historyTile(_ c: ClosetItem) -> some View {
        let isCurrent = c.id == current.id
        let u = c.id.flatMap { chainUrls[$0] } ?? c.image_url
        return Button {
            if !isCurrent { current = c }
        } label: {
            ZStack {
                Palette.card
                if let s = u, let url = URL(string: s) {
                    AsyncImage(url: url) { phase in
                        switch phase {
                        case .success(let img): img.resizable().scaledToFill()
                        default: Color.clear
                        }
                    }
                }
            }
            .frame(width: 56, height: 56)
            .clipShape(RoundedRectangle(cornerRadius: 8))
            .overlay(RoundedRectangle(cornerRadius: 8)
                .stroke(isCurrent ? Palette.bronze : Palette.hairline,
                        lineWidth: isCurrent ? 2 : 1))
        }
        .buttonStyle(.plain)
    }

    private func loadContext() async {
        guard let id = current.id else { return }
        wornCount = (try? await Repo.shared.tryOnCountForPiece(id)) ?? 0

        // Walk parent chain upward
        var parents: [ClosetItem] = []
        var cursor: ClosetItem? = current
        var visited = Set<String>()
        while let cur = cursor, let pid = cur.parent_id, !visited.contains(pid) {
            visited.insert(pid)
            guard let p = try? await Repo.shared.closetItemById(pid) else { break }
            parents.insert(p, at: 0)
            cursor = p
        }
        let kids = (try? await Repo.shared.closetChildren(parentId: id)) ?? []
        let combined = parents + kids
        chain = combined
        var urls: [String: String] = [:]
        for c in combined {
            guard let cid = c.id else { continue }
            if let direct = c.image_url, !direct.isEmpty { urls[cid] = direct; continue }
            guard let path = c.image_path else { continue }
            if let signed = try? await Repo.shared.signedClosetUrl(path) { urls[cid] = signed }
        }
        chainUrls = urls
    }
}
